### PR TITLE
Add code for agglomerative clustering

### PR DESCRIPTION
**Pull request by mathursanjiv**
New clustering algorithm for agglomerative clustering in new files.
Put under option - default off

Past comments from below PR resolved:
https://github.com/The-OpenROAD-Project-private/OpenROAD/pull/409

### CODE REVIEW

File: src/TritonCTS/src/SinkAgglDriver.cpp [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+//   nclust = number of clusters
+// Output arguments:
+//   labels = allocated integer array of size n for result
+//
+void SinkAgglClustering::cutreeK(int n, const int* merge, int nclust, int* labels) {
+
+  int k,m1,m2,j,l;
+
+  if (nclust > n || nclust < 2) {
+    for (j=0; j<n; j++) labels[j] = 0;
```
Comment (maliberty) [inline src/TritonCTS/src/SinkAgglDriver.cpp:72]: {} around the body even if one line.

File: src/TritonCTS/src/SinkAgglDriver.cpp [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+  for (double const * DD=D; DD!=D+(static_cast<std::ptrdiff_t>(N)*(N-1)>>1);
+       ++DD) {
+#if HAVE_DIAGNOSTIC
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wfloat-equal"
+#endif
+
+#if HAVE_DIAGNOSTIC
+#pragma GCC diagnostic pop
+#endif
```
Comment (maliberty) [inline src/TritonCTS/src/SinkAgglDriver.cpp:258]: Why is this here?

**Review by maliberty — COMMENTED** [review]
This code is not really ready for review.  It needs style updating and documenting to make it reviewable for algorithmic correctness.

File: src/TritonCTS/src/SinkAgglDriver.cpp [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+// that the cluster result is split into nclust clusters.
+//
+// Input arguments:
+//   n      = number of observables
+//   merge  = clustering result in R format
+//   nclust = number of clusters
+// Output arguments:
+//   labels = allocated integer array of size n for result
+//
+void SinkAgglClustering::cutreeK(int n, const int* merge, int nclust, int* labels) {
```
Comment (QuantamHD) [inline src/TritonCTS/src/SinkAgglDriver.cpp:67]: Recommend removing all abbreviations from this PR. See Google Style guide on the abbreviations https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

So for example in this signature I would want something closer to

```c++
void SinkAgglClustering::cutreeK(int observableCount, const int* clusterResultsRFormat, int clusterCount, int* labels) {
```

All names should be changed to comply with the no abbreviation rule.

Comment (QuantamHD) [inline src/TritonCTS/src/SinkAgglDriver.cpp:67]: can const int* merge become a std::vector instead?

File: src/TritonCTS/src/SinkAgglDriver.cpp [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+
+  int k,m1,m2,j,l;
+
+  if (nclust > n || nclust < 2) {
+    for (j=0; j<n; j++) labels[j] = 0;
+    return;
+  }
+
+  // assign to each observable the number of its last merge step
+  // beware: indices of observables in merge start at 1 (R convention)
```
[the remark this answers is not in the document]
Comment (QuantamHD) [reply src/TritonCTS/src/SinkAgglDriver.cpp:77]: Agree with Matt here. This makes the code very hard to maintain for others outside the initial authorship group. We should strive for maximum readability rather than minimum code size.

File: src/TritonCTS/src/SinkAgglDriver.cpp [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+//             Result follows R hclust convention:
+//              - observabe indices start with one
+//              - merge[i][] contains the merged nodes in step i
+//              - merge[i][j] is negative when the node is an atom
+//   height  = allocated (n-1) array with distances at each merge step
+// Return code:
+//   0 = ok
+//   1 = invalid method
+//
+int SinkAgglClustering::hclustFast(int n, double* distmat, int method,
```
Comment (QuantamHD) [inline src/TritonCTS/src/SinkAgglDriver.cpp:183]: Would recommend changing this to return an enum instead. Something like

```c++
enum class ClusterFastResult {
  OK,
  INVALID_METHOD
}
```

File: src/TritonCTS/src/SinkAgglDriver.cpp [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+  generateRDendrogram<false>(merge, height, order, Z2, n);
+  
+
+  delete[] order; // only needed for visualization
+  
+  
+  return 0;
+}
+
+// Indexing functions
```
Comment (QuantamHD) [inline src/TritonCTS/src/SinkAgglDriver.cpp:207]: Can you create an iterator instead of a macro to perform this upper triangular indexing?

File: src/TritonCTS/src/SinkClustering.cpp [hunk 7]
```diff
@@ -167,6 +168,221 @@ void SinkClustering::findBestMatching()
   }
 }
 
+
+void SinkClustering::runAgglomerative(unsigned groupSize, float maxDiameter, cts::DBU scaleFactor)
```
Comment (QuantamHD) [inline src/TritonCTS/src/SinkClustering.cpp:172]: I believe this method could use a significant refactor. As a suggestion I think encapsulating a few of the loop bodies into helper functions would help readers get context on the intent of the code.

File: src/TritonCTS/src/SinkClustering.cpp [hunk 8]
```diff
@@ -167,6 +168,221 @@ void SinkClustering::findBestMatching()
...
+
+    // Empty Vector
+    pointsToCluster.clear();
+
+    // Agglomerative for each cluster..
+    for(unsigned pointCounter=0; pointCounter<thetaSolution[clusterCounter].size(); ++pointCounter){
+      pointsToCluster.push_back(thetaSolution[clusterCounter][pointCounter]);
+    }
+
+    // Keep Running Till All Cluster Size <= GroupSize and Cluster Diameter <= _maxDiameter
```
Comment (QuantamHD) [inline src/TritonCTS/src/SinkClustering.cpp:223]: This comment is helpful, but it would also be helpful to understand what we're doing in this while loop and how it helps us achieve our goal.

File: src/TritonCTS/src/SinkClustering.cpp [hunk 9]
```diff
@@ -167,6 +168,221 @@ void SinkClustering::findBestMatching()
...
+    while(true)
```
Comment (QuantamHD) [inline src/TritonCTS/src/SinkClustering.cpp:224]: Is it possible to define the loop condition in the while loop itself rather than relying on breaks to exit the loop. 


while(true) is a pretty scary proposition as well, if there's ever a bug introduced it will hang the entire program. Is there some sanity check we can include that will help escape this loop if it enters an infinite loop.

File: src/TritonCTS/test/simple_test_clustered_aggl.tcl [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+set_wire_rc -clock -layer metal5
+
+clock_tree_synthesis -root_buf CLKBUF_X3 \
+                     -buf_list CLKBUF_X3 \
+                     -wire_unit 20 \
+                     -post_cts_disable \
+                     -sink_clustering_enable \
+                     -distance_between_buffers 100 \
+                     -sink_clustering_size 10 \
+                     -sink_clustering_max_diameter 60 \
```
Comment (QuantamHD) [inline src/TritonCTS/test/simple_test_clustered_aggl.tcl:80]: How were these parameters chosen? Since this is using a new sink clustering algorithm would it be possible to look at the design and choose reasonable parameters instead of relying on user expertise?

File: src/TritonCTS/src/SinkAgglDriver.cpp [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+      queue[idx].pos = pos;
+      queue[idx].node = child-1;
+      ++idx;
+    }
+  } while (idx>0);
+}
+
+#define size_(r_) ( ((r_<N) ? 1 : nodeSize[r_-N]) )
+
+template <const bool sorted>
```
Comment (QuantamHD) [inline src/TritonCTS/src/SinkAgglDriver.cpp:396]: This should become an enum rather than a bool. 

```c++
enum DendrogramOrder {
  SORTED,
  UNSORTED,
}
```

**Review by QuantamHD — CHANGES_REQUESTED** [review]

File: src/TritonCTS/src/SinkAgglDriver.h [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+
+/*
+  Lookup function for a union-find data structure.
+
+  The function finds the root of idx by going iteratively through all
+  parent elements until a root is found. An element i is a root if
+  nodes[i] is zero. To make subsequent searches faster, the entry for
+  idx and all its parents is updated with the root element.
+ */
+class unionFind {
```
Comment (QuantamHD) [inline src/TritonCTS/src/SinkAgglDriver.h:134]: Boost appears to support a disjoint set data structure would it work for this use case?

**Comment by mathursanjiv** [comment]
@shrey-bansal please check the new comments.

**Comment by maliberty** [comment]
@mathursanjiv is this still live?  Its been 20 days with no update.

File: src/TritonCTS/src/SinkAgglDriver.cpp (hunk 5 above)
Comment (jjcherry56) [reply src/TritonCTS/src/SinkAgglDriver.cpp:183]: seems like a simple bool that indicates failure is sufficient

File: src/TritonCTS/src/SinkClustering.cpp (hunk 8 above)
Comment (jjcherry56) [reply src/TritonCTS/src/SinkClustering.cpp:223]: And Most English Sentences Do No Capitalize Each Word

**Comment by mathursanjiv** [comment]
> @mathursanjiv is this still live? Its been 20 days with no update.

Well the developer is now not working on OpenROAD. I'll try to fix code for feedback over next few days.
